Add mandatory argument to RunSingularity to specify binary to use

mem opened the pull request:
Add mandatory argument to RunSingularity to specify binary to use

RunSingularity is using "singularity" as the command it wants to run,
which means it's picking up singularity from the user's PATH, which
might or might not correspond to the singularity binary that matches the
code to be tested (e.g. mconfig --prefix=/usr/local/test, and then run
run the E2E suite without adding /usr/local/test/bin to the path).

Fix this by adding a cmdPath argument to RunSingularity which specifies
where the singularity binary is located. This plays nice with the
existing code that records the location of this binary in the
E2E_CMD_PATH environment variable. It's the responsability of the test
calling RunSingularity to pass the correct value derived from this
environment variable.

This is replacing the existing "name" argument, which is mandatory, but
it's only used if a subtest run is requested, which is optional. Convert
this into an option "WithSubtest", which takes the name of the subtest as
an argument and is self-documenting.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Commit:
Add e2e.AsSubtest to request running singularity as a subtest

Looking at the code, the "name" argument is mandatory, but it's only
used if a subtest run is requested, which is optional. Convert this into
an option "AsSubtest", which takes the name of the subtest as an
argument.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

mem commented on e2e/internal/e2e/singularitycmd.go:
```diff
@@ -109,7 +109,7 @@ type singularityCmd struct {
 	dir        string
 	privileged bool
-	subTest    bool
+	noSubTest  bool
```
no, scratch this. The way the `RunSingularity` function works requires a name for the subtest, even if the subtest is optional. I'll change that instead so that it makes sense.

jscook2345 replied:
Sounds good. Ping me when you'd like the review.

mem replied:
thanks @jscook2345 	

I had to change more than I had initially expected. There are two commits in this PR now: the first one removes the "name" argument, as mentioned in the cover letter. The second one adds a "cmdPath" argument, which is the path to the singularity binary, as obtained from the E2E_CMD_PATH environment variable.

Commit:
Add mandatory argument to RunSingularity to specify binary to use

RunSingularity is using "singularity" as the command it wants to run,
which means it's picking up singularity from the user's PATH, which
might or might not correspond to the singularity binary that matches the
code to be tested (e.g. mconfig --prefix=/usr/local/test, and then run
run the E2E suite without adding /usr/local/test/bin to the path).

Fix this by adding a cmdPath argument to RunSingularity which specifies
where the singularity binary is located. This plays nice with the
existing code that records the location of this binary in the
E2E_CMD_PATH environment variable. It's the responsability of the test
calling RunSingularity to pass the correct value derived from this
environment variable.

Add a helper method to TestEnv called RunSingularity that does *not*
take the cmdPath argument. Since TestEnv already has a CmdPath field,
use that to call the standalone RunSingularity function. This obscures
what's going on, but reduces the chances of calling the wrong binary.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

cclerget commented on e2e/actions/actions.go:
```diff
@@ -65,7 +65,8 @@ func actionRun(t *testing.T) {
 			t,
-			tt.name,
+			testenv.CmdPath,
+			e2e.WithSubtest(tt.name),
```
I like that with name argument, just it would be more expressive with `AsSubTest(name)`

mem replied:
will do

cclerget commented on e2e/actions/actions.go:
```diff
@@ -65,7 +65,8 @@ func actionRun(t *testing.T) {
 		e2e.RunSingularity(
 			t,
-			tt.name,
+			testenv.CmdPath,
```
We should get rid of `testenv.CmdPath`, because it's a superficial argument and instead we could append `filepath.Join(buildcfg.LIBEXECDIR, "singularity/bin")` to `PATH` in `Run` function: https://github.com/sylabs/singularity/blob/master/e2e/suite.go#L38 and just restore `singularity` value in `RunSingularity`

jscook2345 replied:
@cclerget 

I think the argument is that we may not run the installed singularity binary each time. I think that makes sense to make as an input to testing.

cclerget replied:
After slack discussion, this argument is not required and instead we could set `PATH` conditionally to handle case for a different binary locations (pro release tests ...) :

```
path := os.Getenv("PATH")
if path == "" {
    path = "/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin"
}
// to customize path for both singularity and external binaries
// used in e2e tests
e2ePath := os.Getenv("E2E_PREPEND_PATH")
if e2ePath == "" {
    e2ePath = buildcfg.BINDIR
}
os.Setenv("PATH", fmt.Sprintf("%s:%s", e2ePath, path))
```

mem replied:
No, messing with PATH to get the "right" value is the wrong approach here.

Keep in mind that this is a test and the thing being tested is singularity: the path to the singularity binary under test is an input to the test. That binary under test should be capable of finding its components (confguration file, starter binaries, etc) on its own. How the top level test harness obtains the location of this binary should not be a concern of the test.

I suspect the problem you are trying to work around of is the fact that we are passing a *testing.T argument to the test functions, when we probably should pass something like a *e2e.T which embeds (or includes a field that is) a *testing.T. This *e2e.T would contain the information that we want to pass from the top level test harness down to the tests. If we want to do discovery of test functions, it's probably a good idea to define a single type for the functions that implement tests (just like the testing package does).

All that said, what I'm trying to fix in this PR is the incorrect assumption that the singularity binary under test is the first singularity program found in PATH because that's messing with what I actually want to do.

cclerget replied:
There is confusion, the snippet above is intended to be in `e2e/suite.go` not in `RunSingularity`, then you could just restore value to `"singularity"` in `RunSingularity` and `LookPath` in `exec.Command` will does its job by finding the right singularity binary based on `PATH` initialization above and would avoid to pass this argument everytime. Now if you really need to pass a custom singularity binary, you could add a function like `WithBinary(path)` to `RunSingularity`

mem replied:
No, I understood the suggestion correctly. I still say it's going the wrong way about it: it is up to the test harness to locate the binary that is going to be tested. singularity, for better or worse, is built in such a way that it will look in specific paths for the binaries it needs. From the point of view of the E2E tests, you must install singularity (as in `make install`) and from there singularity itself will locate its stuff. The E2E harness is using buildcfg to locate the install location of the singularity binary, so it's testing the binary that matches the current build. That does not need to be that way, but it is that way right now. By looking for singularity in the PATH, as the current code does, it is making a strong and wrong assumption that whatever singularity binary is found that way will match the singularity binary from the current build. In general, that's false.

cclerget replied:
My bad, I edited code snippet above to use the right `buildcfg` path value.

I'm just pointing that we can get rid of this useless argument for `RunSingularity` ... I suggested to initialize `PATH` like above because we can put a custom path first in `PATH` where `LookPath` during call to `exec.Command("singularity" ...)` in `RunSingularity` will consider the first singularity binary found as the one to execute for e2e tests, this doesn't differ of what was done previously with `cmdPath := filepath.Join(buildcfg.BINDIR, "singularity")`.

If you are not happy with `PATH`, this is not a problem, just back to the original issue and the thing I pointed out in my first comment is: why passing singularity binary path as an argument of `RunSingularity` ? which always has the same value and which could involve more work in case of a refactoring, when `exec.Command` in `RunSingularity` could use an initialized variable directly like `exec.Command(e2e.SingularityPath, ...)` or `PATH` or whatever ...

And like I said, if a test require to use another singularity binary for some reasons but I can't figure out a use case for that, there is the possibility to use something like `WithBinary(path)` as a `SingularityCmdOp` argument for `RunSingularity`.

GodloveD replied:
If I understand, @cclerget is trying to make the following point:

Adding the `cmdPath` argument to the `RunSingularity()` function violates the principle of DRY (Don't repeat yourself).  There are ~50 instances of the `RunSingularity()` function that need to be altered in this PR after the argument is added.  And in all 50 calls, the variable is going to be the same right?  Seems tedious and error prone.  So why not just add something like this _into the `RunSingularity()` function itself._
```
cmdPath := filepath.Join(buildcfg.BINDIR, "singularity")
```
And then use it there once instead of setting it outside the function and passing it in 50 times?  (Or you could also just make it a constant in the `e2e` package and import it into the `RunSingularity()` function or whatever.  The point remains the same.)

Is that basically what you are trying to say @cclerget or am I totally misunderstanding your point?

cclerget replied:
Yes that's the point @GodloveD

cclerget commented on e2e/suite.go:
```diff
@@ -102,8 +102,8 @@ func Run(t *testing.T) {
 	// Start registry for tests
-	singularitye2e.PrepRegistry(t, name)
-	defer singularitye2e.KillRegistry(t)
+	singularitye2e.PrepRegistry(t, cmdPath, name)
```
another reason to get rid of `cmdpath`

mem replied:
sorry?

jscook2345 reviewed (approved):
LGTM

cclerget commented on e2e/env/env.go:
```diff
@@ -90,9 +90,8 @@ func (c *ctx) singularityEnv(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest` call

cclerget commented on e2e/imgbuild/imgbuild.go:
```diff
@@ -57,9 +57,8 @@ func (c *imgBuildTests) buildFrom(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/imgbuild/imgbuild.go:
```diff
@@ -134,9 +133,8 @@ func (c *imgBuildTests) nonRootBuild(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/imgbuild/imgbuild.go:
```diff
@@ -177,9 +175,8 @@ func (c *imgBuildTests) buildLocalImage(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"test-sandbox",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/imgbuild/imgbuild.go:
```diff
@@ -424,9 +418,8 @@ func (c *imgBuildTests) buildMultiStageDefinition(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/imgbuild/imgbuild.go:
```diff
@@ -722,9 +715,8 @@ func (c *imgBuildTests) buildDefinition(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance.go:
```diff
@@ -34,9 +34,8 @@ func (c *ctx) testBasicEchoServer(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		instanceName,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance.go:
```diff
@@ -61,9 +60,8 @@ func (c *ctx) testCreateManyInstances(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			instanceName,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance.go:
```diff
@@ -106,9 +104,8 @@ func (c *ctx) testBasicOptions(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		instanceName,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance.go:
```diff
@@ -156,9 +153,8 @@ func (c *ctx) testContain(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		instanceName,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance.go:
```diff
@@ -212,9 +208,8 @@ func (c *ctx) testInstanceFromURI(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			i.name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance_utils.go:
```diff
@@ -31,9 +31,8 @@ type instanceList struct {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"List",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance_utils.go:
```diff
@@ -53,9 +52,8 @@ func (c *ctx) stopInstance(t *testing.T, instance string, stopArgs ...string) (s
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"Stop",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance_utils.go:
```diff
@@ -78,9 +76,8 @@ func (c *ctx) execInstance(t *testing.T, instance string, execArgs ...string) (s
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/instance/instance_utils.go:
```diff
@@ -106,9 +103,8 @@ func (c *ctx) expectedNumberOfInstances(t *testing.T, n int) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"GetNumberOfInstances",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/internal/e2e/env.go:
```diff
@@ -14,3 +16,11 @@ type TestEnv struct {
+// RunSingularity function, ensuring that RunSingularity gets called
+// with the correct singularity path as specified by the test
+// environment.
+func (env TestEnv) RunSingularity(t *testing.T, cmdOps ...SingularityCmdOp) {
```
I left that here as a reminder (or it's up to you if you want to tackle it in this one), to avoid confusion between different function prototypes and also for consistency with other helpers in `e2e/internal` (to pass a `TestEnv` argument instead of currently by example for a function pass a `cmdPath` and for another pass `TestEnv`), I would suggest to rename `RunSingularity(t *testing.T, cmdPath string, cmdOps ...SingularityCmdOp)` in `e2e/internal/e2e/singularitycmd.go` to `runSingularity(t *testing.T, cmdPath string, cmdOps ...SingularityCmdOp)` for internal use only by `TestEnv`, so each helpers using `RunSingularity` would take a `TestEnv` argument

cclerget commented on e2e/internal/e2e/imageverify.go:
```diff
@@ -75,7 +75,8 @@ func ImageVerify(t *testing.T, cmdPath string, imagePath string) {
 		RunSingularity(
 			t,
-			tt.name,
+			cmdPath,
```
what I was saying about helpers and `cmdPath` or `TestEnv`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -37,7 +37,7 @@ func checkOciState(t *testing.T, containerID string, state string) {
 
 	e2e.RunSingularity(
 		t,
-		"State",
```
missing `AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -52,9 +52,8 @@ func genericOciMount(t *testing.T, c *ctx) (string, func()) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciMount",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -81,9 +80,8 @@ func genericOciMount(t *testing.T, c *ctx) (string, func()) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			"OciUmount",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -140,9 +138,8 @@ func (c *ctx) testOciAttach(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciAttachCreate",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -153,29 +150,27 @@ func (c *ctx) testOciAttach(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciAttachStart",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -153,29 +150,27 @@ func (c *ctx) testOciAttach(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciAttachInteractive",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -186,15 +181,14 @@ func (c *ctx) testOciAttach(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciAttachDelete",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -211,9 +205,8 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicCreate",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -224,29 +217,27 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicStart",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -224,29 +217,27 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciPause",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -256,15 +247,14 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciResume",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicAlreadyStarted",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicExecId",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicKill",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicDelete",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicStateAfterDelete",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicKillAfterDelete",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/oci/oci.go:
```diff
@@ -274,72 +264,65 @@ func (c *ctx) testOciBasic(t *testing.T) {
-	e2e.RunSingularity(
+	c.env.RunSingularity(
 		t,
-		"OciBasicStartAfterDelete",
```
missing `e2e.AsSubtest`

cclerget commented on e2e/version/version.go:
```diff
@@ -38,9 +38,8 @@ func (c *ctx) testSemanticVersion(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest`

cclerget commented on e2e/version/version.go:
```diff
@@ -80,9 +79,8 @@ func (c *ctx) testEqualVersion(t *testing.T) {
-		e2e.RunSingularity(
+		c.env.RunSingularity(
 			t,
-			tt.name,
```
missing `e2e.AsSubtest`
